ext/mysql2/extconf.rb

```diff
@@ -1,188 +1,312 @@
-    libs = `#{mc} --libs`.chomp
-  end
-  exit 1 if $? != 0
-  $INCFLAGS += ' ' + includes
-  $libs = libs + " " + $libs
-  rpath_dir = libs
-else
-  inc, lib = dir_config('mysql', '/usr/local')
-  libs = ['m', 'z', 'socket', 'nsl', 'mygcc']
-  found = false
-  while not find_library('mysqlclient', 'mysql_query', lib, "#{lib}/mysql") do
-    exit 1 if libs.empty?
-    found ||= have_library(libs.shift)
-  end
-
-  asplode("mysql client") unless found
-
-  rpath_dir = lib
-end
 
-if have_header('mysql.h')
-  prefix = nil
-elsif have_header('mysql/mysql.h')
-  prefix = 'mysql'
-else
-  asplode 'mysql.h'
-end
+  def detect_by_mysql_config
+    if mysql_config_path
+      warn  "Using mysql_config at #{mysql_config_path}"
 
-%w{ errmsg.h mysqld_error.h }.each do |h|
-  header = [prefix, h].compact.join '/'
-  asplode h unless have_header h
-end
+      ver = `#{mysql_config_path} --version`.chomp.to_f
+      includes = `#{mysql_config_path} --include`.chomp
+      exit 1 if $? != 0
+      libs = `#{mysql_config_path} --libs_r`.chomp
 
-# These gcc style flags are also supported by clang and xcode compilers,
-# so we'll use a does-it-work test instead of an is-it-gcc test.
-gcc_flags = ' -Wall -funroll-loops'
-if try_link('int main() {return 0;}', gcc_flags)
-  $CFLAGS << gcc_flags
-end
+      # MySQL 5.5 and above already have re-entrant code in libmysqlclient (no _r).
+      if ver >= 5.5 || libs.empty?
+        libs = `#{mysql_config_path} --libs`.chomp
+      end
+      exit 1 if $? != 0
 
-if RUBY_PLATFORM =~ /mswin|mingw/
-  # Build libmysql.a interface link library
-  require 'rake'
-
-  # Build libmysql.a interface link library
-  # Use rake to rebuild only if these files change
-  deffile = File.expand_path('../../../support/libmysql.def', __FILE__)
-  libfile = File.expand_path(File.join(rpath_dir, 'libmysql.lib'))
-  file 'libmysql.a' => [deffile, libfile] do |t|
-    when_writing 'building libmysql.a' do
-      # Ruby kindly shows us where dllwrap is, but that tool does more than we want.
-      # Maybe in the future Ruby could provide RbConfig::CONFIG['DLLTOOL'] directly.
-      dlltool = RbConfig::CONFIG['DLLWRAP'].gsub('dllwrap', 'dlltool')
-      sh dlltool, '--kill-at',
-         '--dllname', 'libmysql.dll',
-         '--output-lib', 'libmysql.a',
-         '--input-def', deffile, libfile
+      $INCFLAGS += ' ' + includes
+      $libs = "#{libs} #{$libs}"
+      [includes, libs]
+    else
+      error 'Cannot find mysql_config' if with_config('mysql-config')
+      nil
     end
   end
 
-  Rake::Task['libmysql.a'].invoke
-  $LOCAL_LIBS << ' ' << 'libmysql.a'
+  def error(message)
+    abort("-----\n#{message}\n-----")
+  end
+
+  def warn(message)
+    super("-----\n#{message}\n-----")
+  end
+
+  def asplode(library)
+    error("#{library} is missing. #{asplode_suggestion}")
+  end
+
+  def asplode_suggestion
+    'Please check your installation of MySQL and try again.'
+  end
+
+  def mysql_config_path
+    nil
```

Comment:
I think this would be good - each subclass would have to call super first, or have different names for the subclass finder I suppose.

The philosophy is that if a user specifies a --with value, use it without trying other options.